src/XtensaOptimize.cpp

```diff
@@ -0,0 +1,2331 @@
+            }
+            const Broadcast *maybe_broadcast = first_arg.as<Broadcast>();
+            if (maybe_broadcast) {
+                return Broadcast::make(maybe_broadcast->value, op->type.lanes());
+            }
+
+            const Ramp *maybe_ramp = first_arg.as<Ramp>();
+            if (maybe_ramp) {
+                return Ramp::make(maybe_ramp->base, maybe_ramp->stride, op->type.lanes());
+            }
+
+            if (first_arg.type().is_bool() && first_arg.type().is_scalar()) {
+                return first_arg;
+            }
+
+            return Call::make(op->type, op->name,
+                              {first_arg, op->args[1]},
+                              Call::PureExtern);
+        }
+        return IRGraphMutator::visit(op);
+    }
+
+    Expr visit(const Shuffle *op) override {
+        if (op->is_slice() && op->slice_stride() == 1 && op->vectors.size() == 1) {
+            Expr mutated = mutate(op->vectors[0]);
+            const Call *maybe_call = mutated.as<Call>();
+            if (maybe_call && maybe_call->name == "halide_xtensa_concat_from_native") {
+                int offset = 0;
+                for (int ix = 0; ix < (int)maybe_call->args.size(); ix++) {
+                    if (offset == op->slice_begin()) {
+                        std::vector<Expr> new_args;
+                        int count = 0;
+                        while (count < op->type.lanes()) {
+                            new_args.push_back(maybe_call->args[ix]);
+                            count += maybe_call->args[ix].type().lanes();
+                            ix++;
+                        }
+                        if (count == op->type.lanes()) {
+                            return Call::make(op->type,
+                                              "halide_xtensa_concat_from_native",
+                                              new_args, Call::PureExtern);
+                        }
+                        break;
+                    }
+                    offset += maybe_call->args[ix].type().lanes();
+                }
+            }
+        }
+
+        return IRGraphMutator::visit(op);
+    }
+
+public:
+    SimplifySliceConcat() = default;
+};
+
+Stmt match_xtensa_patterns(const Stmt &stmt, const Target &target) {
+    const int alignment = target.natural_vector_size<uint8_t>();
+    const int lut_size_in_bytes = 2 * target.natural_vector_size<uint8_t>();
+    Stmt s = OptimizeShuffles(alignment, lut_size_in_bytes).mutate(stmt);
+    s = align_loads(s, alignment, 1);
+    // NOTE(vksnk): CSE seemed to break loop carry
+    // s = common_subexpression_elimination(s);
+
+    // Use at most 16 vector registers for carrying values.
+    // NOTE(vksnk): loop_carry seems to be a little finicky right now
+    // but looks like something we'd definitely want to have, so
+    // need to figure out where it goes wrong.
+    s = loop_carry(s, 16);
+    s = simplify(s);
+    for (int ix = 0; ix < 10; ix++) {
```

Comment:
How was this constant (and the other constants below) decided on for the number of iterations performed?